Issues/4033 reader detail refresh

aerych opened the pull request:
Closes #4033 

This PR introduces an updated detail view for the reader.  The approach includes a switch to Swift for the language and a Storyboard for the UI.  This deprecates the old reader view hierarchy and a follow up PR will delete the old detail vc and its related code. 

New features introduced include:
- a revised UI (cc @shauwnandrews)
- showing/hiding parts of the UI when scrolling on the iPad
- an animated effect when toggling likes 
- on the iPad, the detail follows Apple's recommended readable margins and has an all white background

To Test:
Scenario 1
- Test viewing the detail from the post list. 

Scenario 2
- Test viewing the detail from a push notification

Scenario 3
- Test viewing the detail from a notification the Notifications tab

Scenario 4
- Test viewing the detail from the comments view controller. 
- Confirm that a detail opened this way does not show the comments button. 

Scenario 5 - Features
- Test liking a post. Confirm its animation plays
- Test viewing comments
- Test tapping on a tag to preview the tag topic
- Test tapping on a link to view the link in an in-app browser
- Test tapping on a featured image to view its image detail
- Test tapping on an image in the content to view its image detail
- Test playing a video in the content 
- Test sharing from the nav bar
- Test opening the menu
- Test blocking a site (you should navigate away from the post)
- Test previewing a site by tapping on the header
- Test that the preview site feature is disabled when viewing a post detail for a previewed site topic
- Test scrolling down hides the navbar and the footer view, and scrolling up reveals them again. 
- Test that content is resized correctly when the device changes orientation. 

Example:
![simulator screen shot dec 15 2015 5 32 40 pm](https://cloud.githubusercontent.com/assets/1435271/11827575/c329ec44-a352-11e5-8306-e3730c7f27d2.png)

This feature should be merged with the version 5.10 / 6.0 release (whichever we decide to make the milestone).

Needs review: @jleandroperez

jleandroperez commented on WordPress/Classes/ViewRelated/Notifications/Notifications.storyboard:
```diff
@@ -26,7 +26,7 @@
                         <outlet property="tableHeaderView" destination="Uvo-9e-l6I" id="dxx-mK-msp"/>
                         <segue destination="veA-Pg-QAw" kind="show" identifier="NotificationDetailsViewController" id="7uj-uJ-yZ8"/>
-                        <segue destination="xj3-6N-2Jy" kind="show" identifier="ReaderPostDetailViewController" id="2du-eK-Xkt"/>
+                        <segue destination="xj3-6N-2Jy" kind="show" identifier="ReaderViewController" id="2du-eK-Xkt"/>
```
I think this segue should be named `ReaderDetailViewController` perhaps?

aerych replied:
doH!  Good catch

jleandroperez commented on WordPress/Classes/ViewRelated/Reader/ReaderCardDiscoverAttributionView.swift:
```diff
@@ -43,7 +43,7 @@ import WordPressShared
             configurePostAttribution(contentProvider!)
         } else if contentProvider?.sourceAttributionStyle() == SourceAttributionStyle.Site {
-            configureSiteAttribution(contentProvider!)
+            configureSiteAttribution(contentProvider!, verboseAttribution: false)
```
(Sorry about being nipticky!) how can we verify this one? (perhaps it should live in another PR?)

aerych replied:
Nitpicky is good!  :)

This can be tested via the Discover feed.  Find a post that has a site attribution (says "Visit ___) and tap to view the post (don't click on the attribution).  The detail should show a snippet from the site, and show the attribution again, but this time saying "Visit ___ for more".

jleandroperez commented on WordPress/Classes/ViewRelated/Reader/ReaderDetailViewController.swift:
```diff
@@ -0,0 +1,903 @@
+
+
+    private struct DetailAnalyticsConstants
+    {
```
How did we live before nested entities??

jleandroperez commented on WordPress/Classes/ViewRelated/Reader/ReaderDetailViewController.swift:
```diff
@@ -0,0 +1,903 @@
+    public var shouldHideComments = false
+    private var didBumpStats = false
+    private var didBumpPageViews = false
+    private var footerViewHeightConstraintConstant: CGFloat = 0.0
```
= CGFloat(0.0) ?

jleandroperez commented on WordPress/Classes/ViewRelated/Reader/ReaderDetailViewController.swift:
```diff
@@ -0,0 +1,903 @@
+    }
+
+
+    var isLoaded : Bool {
```
Maybe private?

jleandroperez commented on WordPress/Classes/ViewRelated/Reader/ReaderDetailViewController.swift:
```diff
@@ -0,0 +1,903 @@
+            self.title = postTitle
+        } else {
+            self.title = NSLocalizedString("Post", comment:"Placeholder title for ReaderPostDetails.")
+        }
```
`title = post?.postTitle ?? NSLocalizedString("Post", comment:"Placeholder title for ReaderPostDetails.")` 

(But your version is way more readable!)

aerych replied:
Actually, I like that approach better (its more concise), and if the placeholder string is assigned to its own variable (`let placeholder = ...`) then it would keep things very readable.  Let's do that.

jleandroperez commented on WordPress/Classes/ViewRelated/Reader/ReaderDetailViewController.swift:
```diff
@@ -0,0 +1,903 @@
+    private func configureFeaturedImage() {
+        var url = post!.featuredImageURLForDisplay()
+
+        guard url != nil else {
```
I've never seen `guard` used like this. Interesting!

jleandroperez commented on WordPress/Classes/ViewRelated/Reader/ReaderDetailViewController.swift:
```diff
@@ -0,0 +1,903 @@
+
+extension ReaderDetailViewController : WPRichTextViewDelegate
+{
+
```
Extra space here (sorry!)

jleandroperez commented on WordPress/Classes/ViewRelated/Reader/ReaderDetailViewController.swift:
```diff
@@ -0,0 +1,903 @@
+    }
+
+
+    public func richTextView(richTextView: WPRichTextView!, didReceiveLinkAction linkURL: NSURL!) {
```
If linkURL can never be nil, can we kill the `!` ?. (If it can, line 876 will explode)

aerych replied:
We can totally kill the `!`

jleandroperez commented on WordPress/Classes/ViewRelated/Reader/ReaderDetailViewController.swift:
```diff
@@ -0,0 +1,903 @@
+extension ReaderDetailViewController : UIScrollViewDelegate
+{
+    public func scrollViewWillEndDragging(scrollView: UIScrollView, withVelocity velocity: CGPoint, targetContentOffset: UnsafeMutablePointer<CGPoint>) {
+        if UIDevice.isPad() || footerView.hidden || !isLoaded {
```
Out of curiosity... why the `isPad` check?

aerych replied:
We'll never hide the navbar or the footerview on the iPad, even if the reader is opened in multitasking split view, only on the iPhone.  The rationale is, regardless of model, the iPad always has abundant vertical space.

jleandroperez commented on WordPress/Classes/ViewRelated/Reader/ReaderSpacerView.swift:
```diff
@@ -0,0 +1,21 @@
+        return CGSize(width: space, height: space)
+    }
+
+}
```
Missing newline

jleandroperez commented:
@aerych the new UX is beautiful!. I've found few glitches:

#### Rotation: Text isn't getting re-layed-out
1. Launch the app in portrait mode
2. Open any post's details
3. Rotate the device

As a result, the text will remain capped to the portrait width.

#### Brief Posts
1. Launch the app and open a (short) post
2. Note that the full height of the screen will be white. 

Please, disregard if this is actually intended!

#### ReaderMenu Hidden on scroll down
1. Launch the app and open a (long) post
2. Scroll all the way down, without lifting your finger

The ReaderMenu will get hidden as soon as you scroll down. Just wondering if we actually wanna have that behavior. Suppose you're reading an article, and once you're ready, you'd like to Comment / Like.
(You'd need to scroll upwards just to get the Menu unhidden).

Not sure if this is a bug, might actually be intended!

jleandroperez commented:
Please, disregard the `Brief Posts` comment. Just tested on the iPad... nooow i see where this is going, i like it!

aerych commented:
HIya @jleandroperez! 
Thanks for the review!  I've updated things and the PR is ready for round two. Re: comments left

> Rotation: Text isn't getting re-layed-out

Good catch! This was happening only for posts that were pure text. If there were images and/or links the text would layout correctly on rotation. 

> Brief Posts

Disregarding per your comment :) 

> ReaderMenu Hidden on scroll down

Totally intentional

jleandroperez commented:
@aerych (Still testing!). I accidentally triggered a crash... by opening the iOS Settings (the bottom panel? i was just unlocking Rotation).

I believe this can be solved by just switching `handleApplicationDidBecomeActive` to *public*

```
2015-12-18 10:09:21.962 WordPress[4104:3446264] -[WordPress.ReaderDetailViewController handleApplicationDidBecomeActive:]: unrecognized selector sent to instance 0x12f74e2a0
2015-12-18 10:09:21.963 WordPress[4104:3446264] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[WordPress.ReaderDetailViewController handleApplicationDidBecomeActive:]: unrecognized selector sent to instance 0x12f74e2a0'
*** First throw call stack:
(0x183031900 0x18269ff80 0x18303861c 0x1830355b8 0x182f3968c 0x182fd6fc4 0x182fd67e4 0x182fd6564 0x18303bde4 0x182f170f4 0x183906d2c 0x187dace20 0x187fcad20 0x1845d8b34 0x1839b6018 0x184563704 0x1845d3b84 0x1845d890c 0x187fcb5b4 0x187fcb210 0x1845e7790 0x1845e7b10 0x182fe8efc 0x182fe8990 0x182fe6690 0x182f15680 0x184424088 0x187d8cd90 0x100035888 0x182ab68b8)
libc++abi.dylib: terminating with uncaught exception of type NSException
(lldb) 
```

jleandroperez commented:
@aerych i've just found a thought glitch. I've intended to capture a video, but for some reason, QuickTime stops recording whenever i rotate the device.

This seems to happen on *very specific posts* in which there is no scrolling needed, whenever the navBar is hidden. Steps:

1. Launch the app + open the Reader Details (i'll send you the specific Post via PM)
2. Rotate the device into landscape mode
3. Scroll down so that the navBar gets hidden
4. Rotate again into Portrait mode

As a result, the Reader Details won't be showing the navBar, and apparently, the "Pull from NavigationStack" gesture seems broken. The only way to get back is by tapping the tabBar *Reader* button.

Thanks!

jleandroperez commented:
#### Crash when instantiating from Notifications
1. Receive a 'Mention in post' Notification
2. Open the Notifications tab
3. Tap over the notification

As a result, the app [will crash here](https://cloudup.com/cbMnegt6g6y).
I've patched the Storyboard, it was breaking before getting there (apparently the Storyboard placeholder was a UITableViewController, i've re-done the Reader placeholder). This one seems to be just a sequence glitch!

aerych commented:
Hi @jleandroperez !
Nice bug finds.  I never thought to test the settings panel opening and dismissing. Adding that one to my repertoire. :)  And thanks for sorting out the issue instantiating from Notifications for me. 

> I believe this can be solved by just switching handleApplicationDidBecomeActive to public

I just removed the `private` modifier, since the method doesn't really need to be `public`.

> This seems to happen on very specific posts in which there is no scrolling needed, whenever the navBar is hidden

I knew there were going to be some edge cases we'd need to address. :)  I've changed things so bars will never be hidden unless there is at least twice as much room to scroll as the height of the bars combined. This should ensure there is always room to scroll back up  I've also made the bars unhide whenever rotating from landscape orientation back to portrait.  

Ready for another peek at your leisure sir!

jleandroperez commented:
All fixes confirmed. Awesome job @aerych!!

:shipit:

aerych commented:
Thanks @jleandroperez :)  And thanks again for the assist with the Notifications storyboard

jleandroperez commented:
Anytime Eric, happy to help!
